### DISCUSSION TITLE
Returning values is harder than it could be.

### DISCUSSION THREAD

**Discussion by michaelcheers** [dotnet/csharplang#142] [General]
I find myself writing
```csharp
return result;
```
a lot. So, I think there should be a new feature which lets a variable be a result.
```csharp
static auto Fibonacci (int upTo) //auto is a tuple of all "returning" values, so (int result).
{
    returning int result = 1;
    int lastNum = 0;
    for (int n = 0; n < upTo; n++)
    {
         int oldResult = result;
         result += lastNum;
         lastNum = oldResult;
    }
} 
```
So, because we put returning, it returns it at the end of the function. Also, so if you have to return something somewhere else: you could use
```csharp
return;
```
So, `return;` would return the "returning" values.

**Comment by Joe4evr** [comment]
I can see this being useful when encapsulating an existing `Try*` method. Instead of having to `return success;` in the following example, it could be written as:
```cs
public bool TryAdd(int key, string value)
{
    returning bool success = _encapsulatedDictionary.TryAdd(key, value);
    if (success)
    {
        Console.WriteLine($"Added value: {value}");
    }
}
```

**Comment by iam3yal** [comment]
How about `yield int result = 1;`? :)

**Comment by michaelcheers** [comment]
```csharp
yield int result = 1;
```
is not valid code.
See [msdn](https://msdn.microsoft.com/en-us/library/9k7k7cf0.aspx).
Yield is associated with returning results and going back into the function again.
So, I guess it has the returning multiple results bit.

**Comment by iam3yal** [comment]
@michaelcheers I know it's invalid code today but it can be valid tomorrow though much like your code, also, just because atm it's associated with one thing doesn't mean that tomorrow it can't be associated with another thing. :)

**Comment by michaelcheers** [comment]
@eyalsk, oh I thought you meant that this was feature duplicate, yeah `yield` could be a replacement for `returning.`

**Comment by DavidArno** [comment]
Why not go the whole way and support embedding the entire thing in the for-loop:
````cs
static int Fibonacci(int upTo) =>
    for (var result = 1, var lastNum = 0, n = 0;
         n < upTo || return result;
         oldResult = result, result += lastNum, lastNum = oldResult, n++);
````

Or to put it another way, IMO you are making the code less readable, just to save typing four extra characters.

**Comment by iam3yal** [comment]
I don't mind this feature but I agree with @DavidArno it can make the code less readable, especially for people that are too used to have everything that needs to be returned at the return statement.

**Comment by HaloFour** [comment]
VB used to have a feature similar to this.  Instead of designating a specific variable as the value to be returned it would allow you to assign values to the function name and that value would be returned from the function.

Having been a VB programmer at that time I can vouch that this feature definitely made code more difficult to read.  The calculation that would produce the result could be far removed from the function exit.  Worse, just like this proposal, that feature allowed for the `return` statement to be omitted entirely.  It was common for someone to accidentally set the function name to a value and then completely forget to return, causing what was intended to be a terminal branch in logic to continue execution of the function accidentally.

Fortunately with VB.NET that feature had been removed from the language and now you have to `Return` a value explicitly.

**Comment by gafter** [comment]
@HaloFour No, this feature is still in VB.Net.

**Comment by HaloFour** [comment]
@gafter 

Yes, I just noticed.  I had (blissfully) forgotten the syntax when I tried it again in VB.NET so I was trying to blend it with `Return` which obviously needs the value provided.  For those interested, you'd either just let the function finish or use `Exit Function` explicitly to bail out of the function and it would then return whatever value you happened to assign to that implicit variable.

Either way, from my experience with this feature in VB, I'd argue against implementing it in C#.  It's simple enough to declare a tallying variable if you want one, and you should have to be explicit with when and where your function may exit, particularly if it's supposed to be producing a value.

**Comment by HaloFour** [comment]
As to the other element of the proposal, that is the implicit return type.  In my opinion the method signature represents the contract which is much more important than the method body itself.  Having the return value derived from whatever combination of these `returning`-decorated variables would make it exceptionally hard to see what that method returns at a glance.  It's also not clear at all what should happen if multiple `returning` variables are declared in different exclusive scopes, possibly even with the same name.

```cs
public auto Foo(bool condition) {
    if (condition) {
        returning int result = 123;
    }
    else {
        returning int result = 456;
    }
}
```

Does that method return `(int result)` or `(int result, int result)`?  Neither makes much sense.

Probably the only time I could see allowing for inference of the return value of a method might be when it has an expression body as at least the compiler could determine the type of that expression.

**Comment by DavidArno** [comment]
After posting my comments about "reducing readability just to save a few characters", I felt a twinge of hypocrisy, as I often bang on about "reducing noise" by reducing the amount that must be typed...

Thinking about it though, I've realised there is no hypocrisy. For me, the key is the old adage that "code is written once and read a thousand times" (or some such, I just made those words up). For me, good proposals aren't about "I often find myself having to write...", but "I often find myself having to read...". Proposals for language enhancements that reduce the amount I must read to understand the code are good. Those that simply reduce the amount I must type, yet increase the difficulty of reading the resultant code, are not so good. 

I feel that both the use of `auto` for the return type, and the defining of the return at the start of the method, rather than at the point of return, *increase* the difficulty of reading the resultant code. Therefore I feel this is a "not so good" proposal.

**Comment by michaelcheers** [comment]
@HaloFour, I think that code would be invalid because `returning value` (s) would have to be declared.

**Comment by HaloFour** [comment]
@michaelcheers 

> I think that code would be invalid because `returning value` (s) would have to be declared.

They are being declared, each within a separate block within the method.  Are you saying that these variables have to be declared at the top of the method?  That goes against the C# convention of only declaring variables where they are being used.
